File: ios-native/Sprout/Model/Almanac.swift
import Foundation

/// Статистика сада за период: считается из журнала поливов и сада, как
/// `Score`, а не копится счётчиками. Два рода времени не смешиваются:
/// привычки — поливы по дням, часам и дням недели — в настоящем времени
/// журнала; точность — по остатку воды в миг полива; прогноз — в днях сада,
/// тех же, что на карточках.
struct Almanac: Sendable {
    /// Неделя и месяц — по дням, год — по месяцам; «всё время» — по дням,
    /// пока саду меньше полутора месяцев, потом по месяцам.
    enum Period: String, CaseIterable, Identifiable, Sendable {
        case week, month, year, all

        var id: String { rawValue }

        var title: String {
            switch self {
            case .week: Lang.text("Неделя")
            case .month: Lang.text("Месяц")
            case .year: Lang.text("Год")
            case .all: Lang.text("Всё время")
            }
        }
    }

    /// Столбик графика: день или месяц и поливы в нём.
    struct Bar: Identifiable, Hashable, Sendable {
        var start: Date
        var count: Int

        var id: Date { start }
    }

    /// Сколько воды оставалось в земле, когда поливали. Зоны — те же, что у
    /// тени карточки: без тени — заранее, оранжевая — вовремя, красная — в
    /// последний момент, сухая земля — пересохло.
    struct Aim: Hashable, Sendable {
        enum Zone: Int, CaseIterable, Identifiable, Sendable {
            case early, onTime, lastMoment, dry

            var id: Int { rawValue }

            var title: String {
                switch self {
                case .early: Lang.text("Заранее")
                case .onTime: Lang.text("Вовремя")
                case .lastMoment: Lang.text("В последний момент")
                case .dry: Lang.text("Пересохло")
                }
            }

            /// Что было в земле — подпись под названием.
            var range: String {
                switch self {
                case .early: Lang.text("40% воды и больше")
                case .onTime: Lang.text("20–40% воды")
                case .lastMoment: Lang.text("меньше 20% воды")
                case .dry: Lang.text("земля сухая")
                }
            }
        }

        /// Досуха — меньше процента: столько не видно и на карточке.
        static let dryBelow = 0.01

        private(set) var counts = [0, 0, 0, 0]

        /// Десять корзин по десять процентов: 0–10, …, 90–100.
        private(set) var bins = Array(repeating: 0, count: 10)

        /// Обычный остаток — медиана: один полив «на всякий случай» среднее
        /// сдвинул бы, а медиану нет.
        private(set) var typical: Double?

        var known: Int { counts.reduce(0, +) }

        func count(_ zone: Zone) -> Int { counts[zone.rawValue] }

        func share(_ zone: Zone) -> Double {
            known == 0 ? 0 : Double(count(zone)) / Double(known)
        }

        /// Зона, куда пришлось больше всего поливов; при равенстве — та,
        /// что ближе к «вовремя».
        var usual: Zone? {
            guard known > 0 else { return nil }
            let order: [Zone] = [.onTime, .lastMoment, .early, .dry]
            return order.max { count($0) < count($1) }
        }

        static func zone(_ left: Double) -> Zone {
            switch left {
            case ..<dryBelow: .dry
            case ..<Thirst.alarmBelow: .lastMoment
            case ..<Thirst.warnBelow: .onTime
            default: .early
            }
        }

        init() {}

        /// Записи прежних сборок без остатка воды не считаются: гадать о них
        /// нечем.
        init(_ entries: [Watering]) {
            let lefts = entries.compactMap(\.left).map { min(max($0, 0), 1) }
            for left in lefts {
                counts[Self.zone(left).rawValue] += 1
                // Запас на двоичную дробь: 0.3 × 10 выходит 2.999…
                bins[min(Int(left * 10 + 1e-9), 9)] += 1
            }
            guard !lefts.isEmpty else { return }
            let sorted = lefts.sorted()
            let middle = sorted.count / 2
            typical = sorted.count % 2 == 1 ? sorted[middle]
                : (sorted[middle - 1] + sorted[middle]) / 2
        }
    }

    /// День недели с поливами. Номер — как у календаря: 1 — воскресенье.
    struct Weekday: Identifiable, Hashable, Sendable {
        var number: Int
        var symbol: String
        var name: String
        var count: Int

        var id: Int { number }
    }

    struct RoomLine: Identifiable, Hashable, Sendable {
        var name: String
        var plants: Int
        /// Средняя влажность сейчас; пусто — растений нет.
        var moisture: Double?
        var waterings: Int
        /// Доля поливов вовремя; пусто — судить не по чему.
        var onTime: Double?

        var id: String { name }
    }

    struct PlantLine: Identifiable, Hashable, Sendable {
        var id: Plant.ID
        var name: String
        var species: String
        var room: String
        var moisture: Double
        /// Дней сада до полива — тем же счётом, что на карточке.
        var due: Int
        var waterings: Int
        var total: Int
        var last: Date?
        var typical: Double?
    }

    /// Сад в эту минуту.
    struct Now: Hashable, Sendable {
        var calm = 0
        var warn = 0
        var alarm = 0
        var average: Double?
        /// Влажность каждого — от самого сухого: полоска на обзоре.
        var levels: [Double] = []
        var driest: String?

        var count: Int { calm + warn + alarm }

        /// Доля довольных — кольцо на обзоре.
        var content: Double {
            count == 0 ? 0 : Double(calm) / Double(count)
        }
    }

    /// День сада в прогнозе: кому в этот день понадобится вода.
    struct Ahead: Identifiable, Hashable, Sendable {
        var offset: Int
        var plants: [Plant.ID] = []
        var names: [String] = []

        var id: Int { offset }
        var count: Int { plants.count }
    }

    var period: Period = .week
    var start = Date(timeIntervalSince1970: 0)
    var byMonth = false
    var bars: [Bar] = []
    var total = 0
    /// Тот же срок перед периодом; у «всего времени» сравнивать не с чем.
    var previous: Int?
    /// Дней с поливом и дней в периоде.
    var active = 0
    var length = 1
    var streak = 0
    var best = 0
    var aim = Aim()
    var hours = Array(repeating: 0, count: 24)
    var weekdays: [Weekday] = []
    var rooms: [RoomLine] = []
    var plants: [PlantLine] = []
    var now = Now()
    var ahead: [Ahead] = []

    /// Прогноз — на две недели сада.
    static let horizon = 14

    /// «Всё время» по дням — пока саду меньше стольких дней.
    static let dailyUpTo = 45

    /// Разница с прошлым периодом; пусто — сравнивать не с чем.
    var change: Int? { previous.map { total - $0 } }

    /// Час, в который поливают чаще всего; пусто — поливов не было.
    var peakHour: Int? {
        guard let most = hours.max(), most > 0 else { return nil }
        return hours.firstIndex(of: most)
    }

    var peakDay: Weekday? {
        guard let most = weekdays.map(\.count).max(), most > 0 else {
            return nil
        }
        return weekdays.first { $0.count == most }
    }

    /// Больше всего поливов в одном столбике — потолок графика.
    var tallest: Int { bars.map(\.count).max() ?? 0 }

    /// Календарь и «сейчас» — снаружи: прогон модели не должен зависеть от
    /// часового пояса и дня запуска.
    static func of(_ log: [Watering], rooms: [Room], period: Period,
                   since: Date, now moment: Date = Date(),
                   calendar: Calendar = .current) -> Almanac {
        var book = Almanac()
        book.period = period
        let today = calendar.startOfDay(for: moment)
        let (start, byMonth) = window(period, log: log, since: since,
                                      today: today, calendar: calendar)
        book.start = start
        book.byMonth = byMonth
        let inside = log.filter { $0.when >= start && $0.when <= moment }
        book.total = inside.count
        book.length = max(1, days(from: start, to: today,
                                  calendar: calendar) + 1)
        if let back = before(period, start: start, calendar: calendar) {
            book.previous = log.count { $0.when >= back && $0.when < start }
        }
        book.active = Set(inside.map { calendar.startOfDay(for: $0.when) })
            .count
        let score = Score.of(log, rooms: rooms, now: moment,
                             calendar: calendar)
        book.streak = score.streak
        book.best = score.best
        book.bars = bars(inside, from: start, to: today, byMonth: byMonth,
                         calendar: calendar)
        book.aim = Aim(inside)
        for entry in inside {
            book.hours[calendar.component(.hour, from: entry.when)] += 1
        }
        book.weekdays = weekdays(inside, calendar: calendar)
        book.rooms = rooms.map { room in
            let ids = Set(room.plants.map(\.id))
            let mine = inside.filter { ids.contains($0.plant) }
            let aim = Aim(mine)
            let levels = room.plants.map(\.moisture)
            return RoomLine(
                name: room.name, plants: room.plants.count,
                moisture: levels.isEmpty ? nil
                    : levels.reduce(0, +) / Double(levels.count),
                waterings: mine.count,
                onTime: aim.known == 0 ? nil : aim.share(.onTime))
        }
        let byPlant = Dictionary(grouping: log, by: \.plant)
        book.plants = rooms.flatMap { room in
            room.plants.map { plant in
                let all = byPlant[plant.id] ?? []
                return PlantLine(
                    id: plant.id, name: plant.name, species: plant.species,
                    room: room.name, moisture: plant.moisture,
                    due: plant.daysUntilWatering,
                    waterings: all.count { $0.when >= start
                        && $0.when <= moment },
                    total: all.count,
                    last: all.map(\.when).max(),
                    typical: Aim(all).typical)
            }
        }
        book.now = now(rooms)
        book.ahead = ahead(rooms)
        return book
    }

    /// Начало периода и шаг графика. Неделя — семь дней с сегодняшним, месяц
    /// — тридцать; год — двенадцать месяцев с текущим, чтобы столбики были
    /// целыми месяцами.
    static func window(_ period: Period, log: [Watering], since: Date,
                       today: Date, calendar: Calendar) -> (Date, Bool) {
        switch period {
        case .week:
            return (calendar.date(byAdding: .day, value: -6, to: today)
                        ?? today, false)
        case .month:
            return (calendar.date(byAdding: .day, value: -29, to: today)
                        ?? today, false)
        case .year:
            let month = calendar.dateInterval(of: .month, for: today)?.start
                ?? today
            return (calendar.date(byAdding: .month, value: -11, to: month)
                        ?? month, true)
        case .all:
            let first = ([since] + log.map(\.when)).min() ?? today
            let start = min(calendar.startOfDay(for: first), today)
            guard days(from: start, to: today, calendar: calendar)
                    >= dailyUpTo else { return (start, false) }
            return (calendar.dateInterval(of: .month, for: start)?.start
                        ?? start, true)
        }
    }

    /// Начало прошлого такого же периода.
    private static func before(_ period: Period, start: Date,
                               calendar: Calendar) -> Date? {
        switch period {
        case .week: calendar.date(byAdding: .day, value: -7, to: start)
        case .month: calendar.date(byAdding: .day, value: -30, to: start)
        case .year: calendar.date(byAdding: .month, value: -12, to: start)
        case .all: nil
        }
    }

    /// По календарю, а не делением секунд: сутки бывают в 23 и 25 часов.
    private static func days(from start: Date, to end: Date,
                             calendar: Calendar) -> Int {
        max(0, calendar.dateComponents([.day], from: start, to: end).day ?? 0)
    }

    /// Пустые дни тоже: без них график врал бы о промежутках.
    static func bars(_ entries: [Watering], from start: Date, to today: Date,
                     byMonth: Bool, calendar: Calendar) -> [Bar] {
        let unit: Calendar.Component = byMonth ? .month : .day
        var counts: [Date: Int] = [:]
        for entry in entries {
            let key = calendar.dateInterval(of: unit, for: entry.when)?.start
                ?? calendar.startOfDay(for: entry.when)
            counts[key, default: 0] += 1
        }
        var out: [Bar] = []
        var cursor = calendar.dateInterval(of: unit, for: start)?.start
            ?? start
        // Потолок — на случай сломанного календаря: цикл не должен виснуть.
        while cursor <= today, out.count < 400 {
            out.append(Bar(start: cursor, count: counts[cursor] ?? 0))
            guard let next = calendar.date(byAdding: unit, value: 1,
                                           to: cursor) else { break }
            cursor = next
        }
        return out
    }

    /// С первого дня недели, принятого в стране: у кого понедельник, у кого
    /// воскресенье. Названия — на языке приложения.
    static func weekdays(_ entries: [Watering],
                         calendar: Calendar) -> [Weekday] {
        var counts = Array(repeating: 0, count: 7)
        for entry in entries {
            counts[calendar.component(.weekday, from: entry.when) - 1] += 1
        }
        var namer = Calendar(identifier: .gregorian)
        namer.locale = Lang.locale
        let short = namer.shortStandaloneWeekdaySymbols
        let full = namer.standaloneWeekdaySymbols
        return (0 ..< 7).map { step in
            let index = (calendar.firstWeekday - 1 + step) % 7
            return Weekday(number: index + 1, symbol: short[index],
                           name: full[index], count: counts[index])
        }
    }

    static func now(_ rooms: [Room]) -> Now {
        let plants = rooms.flatMap(\.plants)
        var now = Now()
        for plant in plants {
            switch plant.thirst {
            case .calm: now.calm += 1
            case .warn: now.warn += 1
            case .alarm: now.alarm += 1
            }
        }
        guard !plants.isEmpty else { return now }
        now.average = plants.map(\.moisture).reduce(0, +)
            / Double(plants.count)
        now.levels = plants.map(\.moisture).sorted()
        now.driest = plants.min { $0.moisture < $1.moisture }?.name
        return now
    }

    /// Первый полив — тем же счётом, что на карточке; дальше — через срок,
    /// если поливать вовремя.
    static func ahead(_ rooms: [Room], days: Int = horizon) -> [Ahead] {
        var out = (0 ..< days).map { Ahead(offset: $0) }
        for plant in rooms.flatMap(\.plants) where plant.period > 0 {
            for turn in 0 ..< 1_000 {
                let day = Int((Double(plant.daysUntilWatering)
                               + Double(turn) * plant.period).rounded())
                guard day < days else { break }
                out[day].plants.append(plant.id)
                out[day].names.append(plant.name)
            }
        }
        return out
    }
}

/// Статистика одного растения — для его экрана в статистике.
struct PlantBook: Sendable {
    var total = 0
    var inPeriod = 0
    var last: Date?
    /// За всё время: у одного растения поливов за неделю мало для выводов.
    var aim = Almanac.Aim()
    /// Поливы периода — от старого к новому.
    var recent: [Watering] = []
    /// Три ближайших полива, в днях сада.
    var dues: [Int] = []
    /// Средний промежуток между поливами в настоящем времени, см. `Diary`.
    var average: TimeInterval?

    static let dueCount = 3

    static func of(_ plant: Plant, log: [Watering], period: Almanac.Period,
                   since: Date, now moment: Date = Date(),
                   calendar: Calendar = .current) -> PlantBook {
        var book = PlantBook()
        let mine = log.filter { $0.plant == plant.id }
            .sorted { $0.when < $1.when }
        let today = calendar.startOfDay(for: moment)
        let (start, _) = Almanac.window(period, log: log, since: since,
                                        today: today, calendar: calendar)
        book.total = mine.count
        book.recent = mine.filter { $0.when >= start && $0.when <= moment }
        book.inPeriod = book.recent.count
        book.last = mine.last?.when
        book.aim = Almanac.Aim(mine)
        book.average = Diary.of(log, plant: plant.id).average
        if plant.period > 0 {
            book.dues = (0 ..< dueCount).map { turn in
                Int((Double(plant.daysUntilWatering)
                     + Double(turn) * plant.period).rounded())
            }
        }
        return book
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/model/Almanac.kt
package com.ledro6.sprout.model

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.TextStyle

/**
 * Статистика сада за период: из журнала поливов и сада, как `Score`.
 * Привычки — в настоящем времени журнала; точность — по остатку воды в миг
 * полива; прогноз — в днях сада, тех же, что на карточках.
 */
data class Almanac(
    val period: Period = Period.WEEK,
    val start: LocalDate = LocalDate.of(1970, 1, 1),
    val byMonth: Boolean = false,
    val bars: List<Bar> = emptyList(),
    val total: Int = 0,
    /** Тот же срок перед периодом; у «всего времени» сравнивать не с чем. */
    val previous: Int? = null,
    val active: Int = 0,
    val length: Int = 1,
    val streak: Int = 0,
    val best: Int = 0,
    val aim: Aim = Aim(),
    val hours: List<Int> = List(24) { 0 },
    val weekdays: List<Weekday> = emptyList(),
    val rooms: List<RoomLine> = emptyList(),
    val plants: List<PlantLine> = emptyList(),
    val now: Now = Now(),
    val ahead: List<Ahead> = emptyList(),
    val cells: List<Cell> = emptyList(),
    val records: Records = Records(),
) {
    enum class Period {
        WEEK, MONTH, YEAR, ALL;

        val title: String
            get() = when (this) {
                WEEK -> Lang.text("Неделя")
                MONTH -> Lang.text("Месяц")
                YEAR -> Lang.text("Год")
                ALL -> Lang.text("Всё время")
            }
    }

    /** Столбик графика: день или месяц и поливы в нём. */
    data class Bar(val start: LocalDate, val count: Int)

    /** Сколько воды оставалось в земле, когда поливали. */
    data class Aim(
        val counts: List<Int> = List(4) { 0 },
        /** Десять корзин по десять процентов. */
        val bins: List<Int> = List(10) { 0 },
        /** Обычный остаток — медиана. */
        val typical: Double? = null,
    ) {
        enum class Zone {
            EARLY, ON_TIME, LAST_MOMENT, DRY;

            val title: String
                get() = when (this) {
                    EARLY -> Lang.text("Заранее")
                    ON_TIME -> Lang.text("Вовремя")
                    LAST_MOMENT -> Lang.text("В последний момент")
                    DRY -> Lang.text("Пересохло")
                }

            val range: String
                get() = when (this) {
                    EARLY -> Lang.text("40% воды и больше")
                    ON_TIME -> Lang.text("20–40% воды")
                    LAST_MOMENT -> Lang.text("меньше 20% воды")
                    DRY -> Lang.text("земля сухая")
                }
        }

        val known: Int get() = counts.sum()

        fun count(zone: Zone) = counts[zone.ordinal]

        fun share(zone: Zone): Double = if (known == 0) 0.0 else count(zone).toDouble() / known

        /** Зона, куда пришлось больше всего поливов; при равенстве — ближе к «вовремя». */
        val usual: Zone?
            get() {
                if (known == 0) return null
                val order = listOf(Zone.ON_TIME, Zone.LAST_MOMENT, Zone.EARLY, Zone.DRY)
                var best = order[0]
                for (zone in order.drop(1)) if (count(zone) > count(best)) best = zone
                return best
            }

        companion object {
            const val DRY_BELOW = 0.01

            fun zone(left: Double): Zone = when {
                left < DRY_BELOW -> Zone.DRY
                left < Thirst.ALARM_BELOW -> Zone.LAST_MOMENT
                left < Thirst.WARN_BELOW -> Zone.ON_TIME
                else -> Zone.EARLY
            }

            /** Записи без остатка воды не считаются: гадать о них нечем. */
            fun of(entries: List<Watering>): Aim {
                val lefts = entries.mapNotNull { it.left }.map { it.coerceIn(0.0, 1.0) }
                val counts = IntArray(4)
                val bins = IntArray(10)
                for (left in lefts) {
                    counts[zone(left).ordinal] += 1
                    bins[minOf((left * 10 + 1e-9).toInt(), 9)] += 1
                }
                if (lefts.isEmpty()) return Aim(counts.toList(), bins.toList(), null)
                val sorted = lefts.sorted()
                val middle = sorted.size / 2
                val typical = if (sorted.size % 2 == 1) sorted[middle]
                else (sorted[middle - 1] + sorted[middle]) / 2
                return Aim(counts.toList(), bins.toList(), typical)
            }
        }
    }

    /** День недели с поливами. Номер — как у календаря iOS: 1 — воскресенье. */
    data class Weekday(val number: Int, val symbol: String, val name: String, val count: Int)

    data class RoomLine(
        val name: String,
        val plants: Int,
        val moisture: Double?,
        val waterings: Int,
        val onTime: Double?,
    )

    data class PlantLine(
        val id: String,
        val name: String,
        val species: String,
        val room: String,
        val moisture: Double,
        val due: Int,
        val waterings: Int,
        val total: Int,
        val last: Long?,
        val typical: Double?,
    )

    /** Сад в эту минуту. */
    data class Now(
        val calm: Int = 0,
        val warn: Int = 0,
        val alarm: Int = 0,
        val average: Double? = null,
        val levels: List<Double> = emptyList(),
        val driest: String? = null,
    ) {
        val count: Int get() = calm + warn + alarm
        val content: Double get() = if (count == 0) 0.0 else calm.toDouble() / count
    }

    /** День сада в прогнозе: кому понадобится вода. */
    data class Ahead(val offset: Int, val plants: List<String> = emptyList(), val names: List<String> = emptyList()) {
        val count: Int get() = plants.size
    }

    data class Cell(val day: LocalDate, val count: Int)

    data class Records(
        val busiest: Cell? = null,
        val earliest: Int? = null,
        val latest: Int? = null,
        val longest: Int = 0,
        val total: Int = 0,
    )

    val change: Int? get() = previous?.let { total - it }

    val peakHour: Int?
        get() {
            val most = hours.maxOrNull() ?: return null
            if (most <= 0) return null
            return hours.indexOf(most)
        }

    val peakDay: Weekday?
        get() {
            val most = weekdays.maxOfOrNull { it.count } ?: return null
            if (most <= 0) return null
            return weekdays.firstOrNull { it.count == most }
        }

    val tallest: Int get() = bars.maxOfOrNull { it.count } ?: 0

    companion object {
        const val HORIZON = 14
        const val WEEKS = 16
        const val DAILY_UP_TO = 45

        fun of(log: List<Watering>, rooms: List<Room>, period: Period, since: Long, now: Long, days: Days): Almanac {
            val today = days.day(now)
            val (start, byMonth) = window(period, log, since, today, days)
            val from = days.start(start)
            val inside = log.filter { it.at >= from && it.at <= now }
            val previous = before(period, start)?.let { back ->
                val b = days.start(back)
                log.count { it.at >= b && it.at < from }
            }
            val score = Score.of(log, rooms, now, days)
            val hours = IntArray(24)
            for (entry in inside) hours[days.hour(entry.at)] += 1
            val byPlant = log.groupBy { it.plant }
            return Almanac(
                period = period,
                start = start,
                byMonth = byMonth,
                total = inside.size,
                length = maxOf(1, (days.between(start, today) + 1).toInt()),
                previous = previous,
                active = inside.map { days.day(it.at) }.toSet().size,
                streak = score.streak,
                best = score.best,
                bars = bars(inside, start, today, byMonth, days),
                aim = Aim.of(inside),
                hours = hours.toList(),
                weekdays = weekdays(inside, days),
                rooms = rooms.map { room ->
                    val ids = room.plants.map { it.id }.toSet()
                    val mine = inside.filter { it.plant in ids }
                    val aim = Aim.of(mine)
                    val levels = room.plants.map { it.moisture }
                    RoomLine(
                        name = room.name,
                        plants = room.plants.size,
                        moisture = if (levels.isEmpty()) null else levels.sum() / levels.size,
                        waterings = mine.size,
                        onTime = if (aim.known == 0) null else aim.share(Aim.Zone.ON_TIME),
                    )
                },
                plants = rooms.flatMap { room ->
                    room.plants.map { plant ->
                        val all = byPlant[plant.id] ?: emptyList()
                        PlantLine(
                            id = plant.id, name = plant.name, species = plant.species,
                            room = room.name, moisture = plant.moisture,
                            due = plant.daysUntilWatering,
                            waterings = all.count { it.at >= from && it.at <= now },
                            total = all.size,
                            last = all.maxOfOrNull { it.at },
                            typical = Aim.of(all).typical,
                        )
                    }
                },
                now = now(rooms),
                ahead = ahead(rooms),
                cells = cells(log, today, days),
                records = records(log, score.best, days),
            )
        }

        /** Дни от начала недели шестнадцать недель назад и до сегодня. */
        fun cells(log: List<Watering>, today: LocalDate, days: Days): List<Cell> {
            val back = today.minusDays(7L * (WEEKS - 1))
            val first = days.weekStart(back)
            val from = days.start(first)
            val counts = HashMap<LocalDate, Int>()
            for (entry in log) {
                if (entry.at < from) continue
                val day = days.day(entry.at)
                counts[day] = (counts[day] ?: 0) + 1
            }
            val out = mutableListOf<Cell>()
            var day = first
            while (!day.isAfter(today)) {
                out.add(Cell(day, counts[day] ?: 0))
                day = day.plusDays(1)
            }
            return out
        }

        fun records(log: List<Watering>, longest: Int, days: Days): Records {
            val counts = HashMap<LocalDate, Int>()
            var earliest: Int? = null
            var latest: Int? = null
            for (entry in log) {
                val day = days.day(entry.at)
                counts[day] = (counts[day] ?: 0) + 1
                val at = days.moment(entry.at)
                val minute = at.hour * 60 + at.minute
                earliest = minOf(earliest ?: minute, minute)
                latest = maxOf(latest ?: minute, minute)
            }
            val top = counts.entries.maxWithOrNull(compareBy<Map.Entry<LocalDate, Int>> { it.value }.thenBy { it.key })
            return Records(
                busiest = top?.let { Cell(it.key, it.value) },
                earliest = earliest,
                latest = latest,
                longest = longest,
                total = log.size,
            )
        }

        /** Начало периода и шаг графика. */
        fun window(period: Period, log: List<Watering>, since: Long, today: LocalDate, days: Days): Pair<LocalDate, Boolean> =
            when (period) {
                Period.WEEK -> today.minusDays(6) to false
                Period.MONTH -> today.minusDays(29) to false
                Period.YEAR -> today.withDayOfMonth(1).minusMonths(11) to true
                Period.ALL -> {
                    val first = (listOf(since) + log.map { it.at }).min()
                    val day = days.day(first)
                    val start = if (day.isAfter(today)) today else day
                    if (days.between(start, today) < DAILY_UP_TO) start to false
                    else start.withDayOfMonth(1) to true
                }
            }

        private fun before(period: Period, start: LocalDate): LocalDate? = when (period) {
            Period.WEEK -> start.minusDays(7)
            Period.MONTH -> start.minusDays(30)
            Period.YEAR -> start.minusMonths(12)
            Period.ALL -> null
        }

        /** Пустые дни тоже: без них график врал бы о промежутках. */
        fun bars(entries: List<Watering>, start: LocalDate, today: LocalDate, byMonth: Boolean, days: Days): List<Bar> {
            val counts = HashMap<LocalDate, Int>()
            for (entry in entries) {
                val day = days.day(entry.at)
                val key = if (byMonth) day.withDayOfMonth(1) else day
                counts[key] = (counts[key] ?: 0) + 1
            }
            val out = mutableListOf<Bar>()
            var cursor = if (byMonth) start.withDayOfMonth(1) else start
            while (!cursor.isAfter(today) && out.size < 400) {
                out.add(Bar(cursor, counts[cursor] ?: 0))
                cursor = if (byMonth) cursor.plusMonths(1) else cursor.plusDays(1)
            }
            return out
        }

        /** С первого дня недели, принятого в стране. Названия — на языке приложения. */
        fun weekdays(entries: List<Watering>, days: Days): List<Weekday> {
            val counts = IntArray(7)
            for (entry in entries) counts[sunday(days.moment(entry.at).dayOfWeek) - 1] += 1
            val first = sunday(days.firstDay) - 1
            return (0 until 7).map { step ->
                val index = (first + step) % 7
                val weekday = fromSunday(index + 1)
                Weekday(
                    number = index + 1,
                    symbol = weekday.getDisplayName(TextStyle.SHORT_STANDALONE, Lang.locale),
                    name = weekday.getDisplayName(TextStyle.FULL_STANDALONE, Lang.locale),
                    count = counts[index],
                )
            }
        }

        /** 1 — воскресенье, как у календаря iOS. */
        private fun sunday(day: DayOfWeek): Int = day.value % 7 + 1

        private fun fromSunday(number: Int): DayOfWeek = if (number == 1) DayOfWeek.SUNDAY else DayOfWeek.of(number - 1)

        fun now(rooms: List<Room>): Now {
            val plants = rooms.flatMap { it.plants }
            var calm = 0
            var warn = 0
            var alarm = 0
            for (plant in plants) {
                when (plant.thirst) {
                    Thirst.CALM -> calm += 1
                    Thirst.WARN -> warn += 1
                    Thirst.ALARM -> alarm += 1
                }
            }
            if (plants.isEmpty()) return Now(calm, warn, alarm)
            return Now(
                calm, warn, alarm,
                average = plants.sumOf { it.moisture } / plants.size,
                levels = plants.map { it.moisture }.sorted(),
                driest = plants.minByOrNull { it.moisture }?.name,
            )
        }

        /** Первый полив — тем же счётом, что на карточке; дальше — через срок. */
        fun ahead(rooms: List<Room>, count: Int = HORIZON): List<Ahead> {
            val plants = Array(count) { mutableListOf<String>() }
            val names = Array(count) { mutableListOf<String>() }
            for (plant in rooms.flatMap { it.plants }) {
                if (plant.period <= 0) continue
                for (turn in 0 until 1_000) {
                    val day = (plant.daysUntilWatering + turn * plant.period).roundedInt()
                    if (day >= count) break
                    plants[day].add(plant.id)
                    names[day].add(plant.name)
                }
            }
            return (0 until count).map { Ahead(it, plants[it], names[it]) }
        }
    }
}

/** Статистика одного растения — для его экрана в статистике. */
data class PlantBook(
    val total: Int = 0,
    val inPeriod: Int = 0,
    val last: Long? = null,
    val aim: Almanac.Aim = Almanac.Aim(),
    val recent: List<Watering> = emptyList(),
    val dues: List<Int> = emptyList(),
    val average: Double? = null,
) {
    companion object {
        const val DUE_COUNT = 3

        fun of(plant: Plant, log: List<Watering>, period: Almanac.Period, since: Long, now: Long, days: Days): PlantBook {
            val mine = log.filter { it.plant == plant.id }.sortedBy { it.at }
            val today = days.day(now)
            val (start, _) = Almanac.window(period, log, since, today, days)
            val from = days.start(start)
            val recent = mine.filter { it.at >= from && it.at <= now }
            return PlantBook(
                total = mine.size,
                recent = recent,
                inPeriod = recent.size,
                last = mine.lastOrNull()?.at,
                aim = Almanac.Aim.of(mine),
                average = Diary.of(log, plant.id).average,
                dues = if (plant.period > 0) {
                    (0 until DUE_COUNT).map { (plant.daysUntilWatering + it * plant.period).roundedInt() }
                } else {
                    emptyList()
                },
            )
        }
    }
}
